pkg/resource/api.go

```diff
@@ -204,6 +207,47 @@ func (a *APIUpdatingApplicator) Apply(ctx context.Context, o runtime.Object, ao
 	return errors.Wrap(a.client.Update(ctx, m), "cannot update object")
 }
 
+// ExponentialBackoff runs the given function with retries in exponentially increasing
+// backoff periods with a given maximum.
+type ExponentialBackoff struct {
+	// MaximumBackoff will be used as waiting period if the calculated backoff
+	// exceeds it.
+	MaximumBackoff time.Duration
+}
```

Comment:
https://godoc.org/k8s.io/client-go/util/retry#RetryOnConflict

I went to take a look whether client-go had utilities for exponential backoff and found the above function, which seems at first glance like it could replace most of this PR. I'd prefer to use `RetryOnConflict` rather than an updating applicator if possible.

Reply:
Thanks for finding that, it's exactly what I was looking for.

I got it implemented and it's working. However, it does a retry if and only if the error is a conflict. Under the hood, the function is actually a wrap around `OnError`:
```golang
func RetryOnConflict(backoff wait.Backoff, fn func() error) error {
	return OnError(backoff, errors.IsConflict, fn)
}
```

As you can see, we're able to give an error detection function and I think we should keep retrying for all kinds of api-server errors, like `ServerTimeout` or `ServiceUnavailable` etc. that could happen during either in `Get` or `Update` calls so that we don't have to come back again to this problem in other cases where api-server is down for instance. So, I found an interface in `kerrors` opackage that covers all api-server errors that we can use for filtering. As a result, until the context deadline is exceeded (which is an `error` that doesn't satisfy `APIStatus` interface), it'll keep retrying and the total context timeout is max 1.30min, so we'll do our actual best to store the external name.